File: interactive_engine/executor/store/src/schema/data_type.rs
//
//! Copyright 2020 Alibaba Group Holding Limited.
//! 
//! Licensed under the Apache License, Version 2.0 (the "License");
//! you may not use this file except in compliance with the License.
//! You may obtain a copy of the License at
//! 
//! http://www.apache.org/licenses/LICENSE-2.0
//! 
//! Unless required by applicable law or agreed to in writing, software
//! distributed under the License is distributed on an "AS IS" BASIS,
//! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//! See the License for the specific language governing permissions and
//! limitations under the License.

#![allow(dead_code)]
#[derive(Serialize, Deserialize, Copy, Clone, Debug, Eq, PartialEq, Hash)]
pub enum DataType {
    Bool = 0,
    Char = 1,
    Short = 2,
    Int = 3,
    Long = 4,
    Float = 5,
    Double = 6,
    Bytes = 7,
    String = 8,
    Date = 9,
    Set = 10,
    ListInt = 11,
    ListLong = 12,
    ListFloat = 13,
    ListDouble = 14,
    ListString = 15,
    ListBytes = 16,
    Map = 100,
    Unknown = 1000,
}

impl Default for DataType {
    fn default() -> Self {
        DataType::Unknown
    }
}

impl DataType {
    pub(super) fn new(order: u32, expression: &str) -> Self {
        match order {
            0 => DataType::Bool,
            1 => DataType::Char,
            2 => DataType::Short,
            3 => DataType::Int,
            4 => DataType::Long,
            5 => DataType::Float,
            6 => DataType::Double,
            7 => DataType::Bytes,
            8 => DataType::String,
            9 => DataType::Date,
            10 => DataType::Set,
            11 => {
                match expression.to_lowercase().as_str() {
                    "int" => DataType::ListInt,
                    "long" => DataType::ListLong,
                    "float" => DataType::ListFloat,
                    "double" => DataType::ListDouble,
                    "string" => DataType::ListString,
                    "bytes" => DataType::ListBytes,
                    _ => DataType::Unknown,
                }
            }
            12 => DataType::ListLong,
            13 => DataType::ListFloat,
            14 => DataType::ListDouble,
            15 => DataType::ListString,
            16 => DataType::ListBytes,
            _ => DataType::Unknown,
        }
    }

    pub fn len(&self) -> usize {
        match *self {
            DataType::Bool | DataType::Char => 1,
            DataType::Short => 2,
            DataType::Int | DataType::Float => 4,
            DataType::Long | DataType::Double => 8,
            _ => unimplemented!()
        }
    }

    pub fn is_fixed_len(&self) -> bool {
        match *self {
            DataType::Bool | DataType::Char |
            DataType::Short | DataType::Int |
            DataType::Long | DataType::Float |
            DataType::Double => true,
            _ => false,
        }
    }
}

pub fn parse_str_to_data_type(value: &str) -> Result<DataType, String> {
    match value.to_lowercase().as_str() {
        "bool" => Ok(DataType::Bool),
        "char" => Ok(DataType::Char),
        "short" => Ok(DataType::Short),
        "int" => Ok(DataType::Int),
        "long" => Ok(DataType::Long),
        "float" => Ok(DataType::Float),
        "double" => Ok(DataType::Double),
        "bytes" => Ok(DataType::Bytes),
        "string" => Ok(DataType::String),
        "date" => Ok(DataType::Date),
        v => {
            if v.starts_with("list<") {
                let tmp = &v[5..v.len() - 1];
                let sub_type = DataType::from(tmp);
                match sub_type {
                    DataType::Int => Ok(DataType::ListInt),
                    DataType::Long => Ok(DataType::ListLong),
                    DataType::Float => Ok(DataType::ListFloat),
                    DataType::Double => Ok(DataType::ListDouble),
                    DataType::String => Ok(DataType::ListString),
                    DataType::Bytes => Ok(DataType::ListBytes),
                    _ => Err(format!("data type {} not support yet", v))

                }
            } else if v.starts_with("s<") {
                //todo
                Err(format!("data type {} not support yet", v))
            } else if v.starts_with("m<") {
                //todo
                Err(format!("data type {} not support yet", v))
            } else {
                Err(format!("unknown data type {}", v))
            }
        },
    }
}

impl<'a> From<&'a str> for DataType {
    fn from(value: &str) -> Self {
        match value.to_lowercase().as_str() {
            "bool" => DataType::Bool,
            "char" => DataType::Char,
            "short" => DataType::Short,
            "int" => DataType::Int,
            "long" => DataType::Long,
            "float" => DataType::Float,
            "double" => DataType::Double,
            "bytes" => DataType::Bytes,
            "string" => DataType::String,
            "date" => DataType::Date,
            v => {
                if v.starts_with("list<") {
                    let tmp = &v[5..v.len() - 1];
                    let sub_type = DataType::from(tmp);
                    if sub_type == DataType::Int {
                        DataType::ListInt
                    } else {
                        unimplemented!()
                    }
                } else if v.starts_with("s<") {
                    //todo
                    unimplemented!()
                } else if v.starts_with("m<") {
                    //todo
                    unimplemented!()
                } else {
                    DataType::Unknown
                }
            },
        }
    }
}

File: interactive_engine/sdk-common/src/main/java/com/alibaba/graphscope/compiler/api/schema/DataType.java
/**
 * Copyright 2020 Alibaba Group Holding Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.alibaba.graphscope.compiler.api.schema;

import com.alibaba.graphscope.proto.DataTypePb;
import com.alibaba.graphscope.sdkcommon.meta.InternalDataType;

public enum DataType {
    UNKNOWN(0),
    BOOL(1),
    CHAR(2),
    SHORT(3),
    INT(4),
    LONG(5),
    FLOAT(6),
    DOUBLE(7),
    STRING(8),
    BYTES(9),
    INT_LIST(10),
    LONG_LIST(11),
    FLOAT_LIST(12),
    DOUBLE_LIST(13),
    STRING_LIST(14),
    BYTES_LIST(15),
    DATE(16);

    private final byte type;

    DataType(int type) {
        this.type = (byte) type;
    }

    private static final DataType[] TYPES = DataType.values();

    public static DataType fromId(byte id) {
        if (id < 0 || id >= TYPES.length) {
            throw new IllegalArgumentException("Unknown DataType: [" + id + "]");
        }
        return TYPES[id];
    }

    public static DataType parseProto(DataTypePb pb) {
        return fromId((byte) pb.getNumber());
    }

    public static DataType parseString(String type) {
        String upperType = type.toUpperCase();
        if (upperType.startsWith("LIST<")) {
            // "LIST<value>" -> value
            String typeValue = upperType.substring(5, upperType.length() - 1) + "_LIST";
            return DataType.valueOf(typeValue);
        } else {
            return DataType.valueOf(type);
        }
    }

    public DataTypePb toProto() {
        return DataTypePb.forNumber(type);
    }

    @Override
    public String toString() {
        String dataTypeString = this.name();
        if (dataTypeString.endsWith("_LIST")) {
            String[] dataTypeArray = dataTypeString.split("_");
            dataTypeString = "LIST<" + dataTypeArray[0] + ">";
        }

        return dataTypeString;
    }

    public boolean isFixedLength() {
        switch (this) {
            case BOOL:
            case CHAR:
            case SHORT:
            case INT:
            case LONG:
            case FLOAT:
            case DOUBLE:
                return true;
            default:
                return false;
        }
    }

    public int getTypeLength() {
        switch (this) {
            case BOOL:
                return 1;
            case CHAR:
            case SHORT:
                return 2;
            case INT:
            case FLOAT:
                return 4;
            case LONG:
            case DOUBLE:
                return 8;
            default:
                throw new UnsupportedOperationException("not a fixed length type [" + this + "]");
        }
    }

    public static DataType parseFromDataType(
            com.alibaba.graphscope.sdkcommon.meta.DataType dataType) {
        InternalDataType internalDataType = dataType.getType();
        switch (internalDataType) {
            case BOOL:
                return DataType.BOOL;

            case CHAR:
                return DataType.CHAR;

            case SHORT:
                return DataType.SHORT;

            case INT:
                return DataType.INT;

            case LONG:
                return DataType.LONG;

            case FLOAT:
                return DataType.FLOAT;

            case DOUBLE:
                return DataType.DOUBLE;

            case STRING:
                return DataType.STRING;

            case BYTES:
                return DataType.BYTES;

            case DATE:
                return DataType.STRING;

            case LIST:
                {
                    switch (InternalDataType.valueOf(dataType.getExpression())) {
                        case INT:
                            {
                                return DataType.INT_LIST;
                            }
                        case LONG:
                            {
                                return DataType.LONG_LIST;
                            }
                        case FLOAT:
                            {
                                return DataType.FLOAT_LIST;
                            }
                        case DOUBLE:
                            {
                                return DataType.DOUBLE_LIST;
                            }
                        case STRING:
                            {
                                return DataType.STRING_LIST;
                            }
                        default:
                            {
                                throw new IllegalArgumentException(
                                        "Unsupported property data type " + dataType);
                            }
                    }
                }
            default:
                {
                    throw new IllegalArgumentException(
                            "Unsupported property data type " + dataType);
                }
        }
    }
}
